Use UnixFileDescriptor in IPC and ProcessLauncher

mcatanzaro opened the pull request:
#### 0b420fd88a458a0ba19beeb53b6447b6bfdbbb1c
<pre>
Use UnixFileDescriptor in IPC and ProcessLauncher
<a href="https://bugs.webkit.org/show_bug.cgi?id=280237">https://bugs.webkit.org/show_bug.cgi?id=280237</a>

Reviewed by Carlos Garcia Campos.

Using UnixFileDescriptor wherever possible makes it harder to
accidentally leak file descriptors.

* Source/WebKit/Platform/IPC/Connection.h:
(IPC::Connection::Identifier::Identifier):
(IPC::Connection::Identifier::operator bool const):
* Source/WebKit/Platform/IPC/IPCUtilities.h:
* Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
(IPC::Connection::socketDescriptor const):
(IPC::Connection::platformInitialize):
(IPC::Connection::platformInvalidate):
(IPC::createPlatformConnection):
(IPC::Connection::createConnectionIdentifierPair):
* Source/WebKit/Shared/unix/AuxiliaryProcessMain.cpp:
* Source/WebKit/UIProcess/Launcher/ProcessLauncher.cpp:
* Source/WebKit/UIProcess/Launcher/ProcessLauncher.h:
* Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp:
(WebKit::ProcessLauncher::launchProcess):
(WebKit::ProcessLauncher::platformDestroy): Deleted.
* Source/WebKit/UIProcess/Launcher/playstation/ProcessLauncherPlayStation.cpp:
(WebKit::ProcessLauncher::launchProcess):

Canonical link: <a href="https://commits.webkit.org/288001@main">https://commits.webkit.org/288001@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/77f1a4fd6cd4853354c264b421ef95c2b279025c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81569 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1095 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35517 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/86101 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32563 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1114 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8913 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63654 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21386 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84638 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/788 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74225 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43943 "Failed to checkout and rebase branch from PR 34133") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/688 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/31016 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/72119 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29000 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87537 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8802 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6245 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71979 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8983 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70054 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71212 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17741 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15279 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14192 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8762 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/14290 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8600 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12122 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10407 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
```diff
@@ -96,11 +96,21 @@ class AttachmentInfo {
+void Connection::platformInitialize(Identifier&& identifier)
+{
+#if USE(GLIB)
+    m_socket = adoptGRef(g_socket_new_from_fd(identifier.handle.release(), nullptr));
```
Actually I don't like this, (a) because errors are silently ignored, and (b) especially because on error, the fd is leaked. I will change this to release assert there is no error before landing.

carlosgcampos commented on Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
```diff
@@ -128,7 +128,6 @@ void Connection::platformInitialize(Identifier&& identifier)
 void Connection::platformInvalidate()
 {
 #if USE(GLIB)
-    // In the GLib platform the socket descriptor is owned by GSocket.
```
This is still true, no? we are still passing the ownership to the GSocket

mcatanzaro replied:
Yes it's still true, but it's no longer _conditionally_ true, so the comment to explain the previously-confusing behavior is no longer useful. Notably `m_socketDescriptor` doesn't exist anymore.

This change really should have landed via https://github.com/WebKit/WebKit/pull/35314 but seems I missed it.

mcatanzaro committed:
Use UnixFileDescriptor in IPC and ProcessLauncher
https://bugs.webkit.org/show_bug.cgi?id=280237

Reviewed by Carlos Garcia Campos.

Using UnixFileDescriptor wherever possible makes it harder to
accidentally leak file descriptors.

* Source/WebKit/Platform/IPC/Connection.h:
(IPC::Connection::Identifier::Identifier):
(IPC::Connection::Identifier::operator bool const):
* Source/WebKit/Platform/IPC/IPCUtilities.h:
* Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
(IPC::Connection::socketDescriptor const):
(IPC::Connection::platformInitialize):
(IPC::Connection::platformInvalidate):
(IPC::createPlatformConnection):
(IPC::Connection::createConnectionIdentifierPair):
* Source/WebKit/Shared/unix/AuxiliaryProcessMain.cpp:
* Source/WebKit/UIProcess/Launcher/ProcessLauncher.cpp:
* Source/WebKit/UIProcess/Launcher/ProcessLauncher.h:
* Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp:
(WebKit::ProcessLauncher::launchProcess):
(WebKit::ProcessLauncher::platformDestroy): Deleted.
* Source/WebKit/UIProcess/Launcher/playstation/ProcessLauncherPlayStation.cpp:
(WebKit::ProcessLauncher::launchProcess):

Canonical link: https://commits.webkit.org/288001@main

rkirsling commented on Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
```diff
@@ -128,7 +128,6 @@ void Connection::platformInitialize(Identifier&& identifier)
-    // In the GLib platform the socket descriptor is owned by GSocket.
     m_socket = nullptr;
 #else
     if (m_socketDescriptor != -1)
```
This line is now invalid:
```
C:/BW/PlayStation-Release-Build/build/Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:133:28: error: invalid operands to binary expression ('UnixFileDescriptor' and 'int')
  133 |     if (m_socketDescriptor != -1)
      |         ~~~~~~~~~~~~~~~~~~ ^  ~~
```

mcatanzaro replied:
Sorry, I will land a follow-up

rkirsling replied:
Thanks much!

mcatanzaro replied:
https://github.com/WebKit/WebKit/pull/38142
